### PR TITLE
[CGData][GlobalIsel][Legalizer][DAG][MC][AsmParser][X86][AMX] Use `std::move` to avoid copy

### PR DESCRIPTION
**Pull request by abhishek-kaushik22**

### CODE REVIEW

File: llvm/include/llvm/CGData/StableFunctionMap.h [hunk 1]
```diff
@@ -42,8 +42,8 @@ struct StableFunction {
   StableFunction(stable_hash Hash, const std::string FunctionName,
                  const std::string ModuleName, unsigned InstCount,
```
Comment (RKSimon) [inline llvm/include/llvm/CGData/StableFunctionMap.h:43]: Shouldn't the `const std::string` be `StringRef`?

Comment (abhishek-kaushik22) [reply llvm/include/llvm/CGData/StableFunctionMap.h:43]: I didn't make that change but if it is required, I can fix it here or in another PR.

Comment (RKSimon) [reply llvm/include/llvm/CGData/StableFunctionMap.h:43]: Either is fine

Comment (abhishek-kaushik22) [reply llvm/include/llvm/CGData/StableFunctionMap.h:43]: I've opened a new PR https://github.com/llvm/llvm-project/pull/118749 for the change, I'll remove the move from here as they are not required now.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 2]
```diff
@@ -6824,7 +6824,7 @@ SDValue SelectionDAG::FoldConstantArithmetic(unsigned Opcode, const SDLoc &DL,
             continue;
           }
           APInt Val = cast<ConstantSDNode>(Op)->getAPIntValue();
```
Comment (topperc) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:6826]: Can we change the preceding line to `const APInt &Val` instead? `getAPIntVal` returns a reference

Comment (abhishek-kaushik22) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp]: Done. Thanks.

File: llvm/lib/CodeGen/GlobalISel/LegalityPredicates.cpp [hunk 3]
```diff
@@ -35,7 +35,7 @@ LegalityPredicates::typeInSet(unsigned TypeIdx,
                               std::initializer_list<LLT> TypesInit) {
   SmallVector<LLT, 4> Types = TypesInit;
   return [=](const LegalityQuery &Query) {
-    return llvm::is_contained(Types, Query.Types[TypeIdx]);
+    return llvm::is_contained(std::move(Types), Query.Types[TypeIdx]);
```
Comment (topperc) [inline llvm/lib/CodeGen/GlobalISel/LegalityPredicates.cpp:38]: This seems really weird. is_contained is a template function that takes a universal reference. Why do we need a move here?

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 4]
```diff
@@ -6823,8 +6823,8 @@ SDValue SelectionDAG::FoldConstantArithmetic(unsigned Opcode, const SDLoc &DL,
             ScalarOps.push_back(getUNDEF(OpVT));
             continue;
           }
-          APInt Val = cast<ConstantSDNode>(Op)->getAPIntValue();
-          ScalarOps.push_back(SignExtendInReg(Val, OpVT));
+          const APInt &Val = cast<ConstantSDNode>(Op)->getAPIntValue();
+          ScalarOps.push_back(SignExtendInReg(std::move(Val), OpVT));
```
Comment (topperc) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:6827]: I don't think you need the std::move now.

File: llvm/lib/CodeGen/GlobalISel/LegalityPredicates.cpp (hunk 3 above)
Comment (abhishek-kaushik22) [reply llvm/lib/CodeGen/GlobalISel/LegalityPredicates.cpp]: Sorry, I missed that. I've removed these.

**Review by arsenm — APPROVED** [review]

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**